src/Cardano/Prelude/GHC/Heap/Size.hs

```diff
@@ -0,0 +1,98 @@
+{-# LANGUAGE BangPatterns     #-}
+{-# LANGUAGE MagicHash        #-}
+{-# LANGUAGE UnboxedTuples    #-}
+{-# LANGUAGE UnliftedFFITypes #-}
+
+{-# OPTIONS_GHC -Wno-missing-signatures -Wno-monomorphism-restriction #-}
+
+module Cardano.Prelude.GHC.Heap.Size (
+    CountFailure(..)
+  , computeHeapSize
+  , computeHeapSize'
+  ) where
+
+import Cardano.Prelude.Base
+import qualified Prelude
+
+import Foreign.C.Types
+import GHC.Exts.Heap.ClosureTypes (ClosureType)
+import GHC.Prim
+import GHC.Types
+
+{-------------------------------------------------------------------------------
+  Failure
+-------------------------------------------------------------------------------}
+
+cNO_FAILURE          = 0
+cWORK_LIST_FULL      = 1
+cVISITED_FULL        = 2
+cOUT_OF_MEMORY       = 3
+cUNSUPPORTED_CLOSURE = 4
+
+foreign import ccall "getCountFailure"
+  getCountFailure_ :: IO CUInt
+
+data CountFailure =
+    WorkListFull
+  | VisitedFull
+  | OutOfMemory
+  | UnsupportedClosure ClosureType
+  deriving (Show, Eq)
+
+getCountFailure :: IO (Maybe CountFailure)
+getCountFailure = tr <$> getCountFailure_
+  where
+    tr :: CUInt -> Maybe CountFailure
+    tr err
+      | err == cNO_FAILURE          = Nothing
+      | err == cWORK_LIST_FULL      = Just $ WorkListFull
+      | err == cVISITED_FULL        = Just $ VisitedFull
+      | err == cOUT_OF_MEMORY       = Just $ OutOfMemory
+      | err >= cUNSUPPORTED_CLOSURE = Just $ UnsupportedClosure (typ err)
+      | otherwise = Prelude.error "getCountFailure: impossible"
+
+    typ :: CUInt -> ClosureType
+    typ err = toEnum (fromIntegral (err - cUNSUPPORTED_CLOSURE))
+
+{-------------------------------------------------------------------------------
+  Main API
+-------------------------------------------------------------------------------}
+
+-- | Bind to the C function to count the closure size
+--
+-- It is crucial that this function is marked unsafe, because GHC guarantees
+-- that garbage collection will not occur during an unsafe call
+-- (see <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/ffi-chap.html#guaranteed-call-safety>),
+-- which is crucial for reliable counting (Haskell's GC moves objects around,
+-- and so if a GC occurred during counting, we might end up counting objects
+-- more than once or indeed miss them in the count if they happen to be moved to
+-- a location we marked as already visited. This is also the reason that we
+-- count the size of the object in C rather than in Haskell.
+foreign import ccall unsafe "closureSize"
+  closureSize_ :: CUInt -> CUInt -> Addr# -> IO CULong
+
+computeHeapSize' :: Word -> Word -> a -> IO (Either CountFailure Word64)
```

Comment:
Uh, well, the goal here is to present an API to the user that is free from C specific types; they should not leak out.

Reply:
Ok got it.

Reply:
I should have worded this differently:

> Are you using `Word` and `fromIntegral` here so that you don't need to worry about exactly what an `HTYPE_UNSIGNED_INT` is?

I think I meant something like:

> Are you using `Word` and `fromIntegral` here so that *the caller* doesn't need to worry about exactly what an `HTYPE_UNSIGNED_INT` is?